rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,770 @@
+            &key_schedule_context,
+            self.aead.nonce_len() as u16,
+        )
+        .map_err(expand_err)?;
+
+        Ok(KeySchedule {
+            aead: self.aead,
+            key,
+            base_nonce,
+            seq_num: 0,
+        })
+    }
+
+    fn hpke_suite_id(&self) -> Vec<u8> {
+        // "The value of suite_id depends on where the KDF is used; it is assumed implicit from the
+        // implementation and not passed as a parameter. If used inside a KEM algorithm, suite_id MUST
+        // start with "KEM" and identify this KEM algorithm; if used in the remainder of HPKE, it MUST
+        // start with "HPKE" and identify the entire ciphersuite in use."
+
+        const PREFIX: &[u8] = b"HPKE";
+
+        let kem_id = u16::from(self.suite.kem).to_be_bytes();
+        let kdf_id = u16::from(self.suite.sym.kdf_id).to_be_bytes();
+        let aead_id = u16::from(self.suite.sym.aead_id).to_be_bytes();
+
+        [PREFIX, &kem_id, &kdf_id, &aead_id].concat()
+    }
+}
+
+impl Hpke for &Suite {
+    fn seal(
+        &mut self,
+        info: &[u8],
+        aad: &[u8],
+        plaintext: &[u8],
+        pub_key: &HpkePublicKey,
+    ) -> Result<(EncapsulatedSecret, Vec<u8>), Error> {
+        let (encap, mut sealer) = Sealer::new(self, info, pub_key)?;
+        Ok((encap, sealer.seal(aad, plaintext)?))
+    }
+
+    fn setup_sealer(
+        &mut self,
+        info: &[u8],
+        pub_key: &HpkePublicKey,
+    ) -> Result<(EncapsulatedSecret, Box<dyn HpkeSealer + 'static>), Error> {
+        let (encap, sealer) = Sealer::new(self, info, pub_key)?;
+        Ok((encap, Box::new(sealer)))
+    }
+
+    fn open(
+        &mut self,
+        enc: &EncapsulatedSecret,
+        info: &[u8],
+        aad: &[u8],
+        ciphertext: &[u8],
+        secret_key: &HpkePrivateKey,
+    ) -> Result<Vec<u8>, Error> {
+        Opener::new(self, enc, info, secret_key)?.open(aad, ciphertext)
+    }
+
+    fn setup_opener(
+        &mut self,
+        enc: &EncapsulatedSecret,
+        info: &[u8],
+        secret_key: &HpkePrivateKey,
+    ) -> Result<Box<dyn HpkeOpener + 'static>, Error> {
+        Ok(Box::new(Opener::new(self, enc, info, secret_key)?))
+    }
+}
+
+impl Debug for Suite {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        self.suite.fmt(f)
+    }
+}
+
+/// Adapts a [KeySchedule] for the role of a [HpkeSealer].
+struct Sealer {
+    key_schedule: KeySchedule,
+}
+
+impl Sealer {
+    /// See RFC 9180 §5.1.1 "Encryption to a Public Key"[^0].
+    ///
+    /// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-5.1.1
+    fn new(
+        suite: &Suite,
+        info: &[u8],
+        pub_key: &HpkePublicKey,
+    ) -> Result<(EncapsulatedSecret, Self), Error> {
+        // def SetupBaseS(pkR, info):
+        //   shared_secret, enc = Encap(pkR)
+        //   return enc, KeyScheduleS(mode_base, shared_secret, info,
+        //                            default_psk, default_psk_id)
```

Comment:
Throughout I've been putting the pseudo-code from the RFC up-front where applicable. I found this helpful while writing it, and thought it might be useful for reviewers.

It's also a bit verbose, and it isn't always 1:1 (e.g. variable names, AEAD Seal being in-lined, etc).

Should these comments be removed prior to merge or do we like them?

Reply:
I like these comments. I would be tempted to intermingle with them with the code they apply to, eg:

```rust
        // def SetupBaseS(pkR, info):
        //   shared_secret, enc = Encap(pkR)
        let (shared_secret, enc) = suite.dh_kem.encap(pub_key)?;

        //   return enc, KeyScheduleS(mode_base, shared_secret, info,
        //                            default_psk, default_psk_id)
        let key_schedule = suite.key_schedule(shared_secret, info)?;
        Ok((enc, Self { key_schedule }))
```

Then, if there is a line in the standard pseudocode that is happening but non-obvious (like, below, `self.IncrementSeq()`) that should be clearer.